ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/ChainDataPruner.java

```diff
@@ -0,0 +1,188 @@
+
+  private static final Bytes PRUNING_MARK_KEY =
+      Bytes.wrap("blockNumberTail".getBytes(StandardCharsets.UTF_8));
+
+  private static final Bytes VARIABLES_PREFIX = Bytes.of(1);
+  private static final Bytes FORK_BLOCKS_PREFIX = Bytes.of(2);
+
+  // TODO: cleanup - pruningEnabled will not be required after
+  // https://github.com/hyperledger/besu/pull/4703 is merged.
+  private static boolean pruningEnabled = false;
+  private final BlockchainStorage blockchainStorage;
+  private final KeyValueStorage prunerStorage;
+  private final long blocksToRetain;
+
+  private final long pruningFrequency;
+
+  public ChainDataPruner(
+      final BlockchainStorage blockchainStorage,
+      final KeyValueStorage prunerStorage,
+      final long blocksToRetain,
+      final long pruningFrequency) {
+    this.blockchainStorage = blockchainStorage;
+    this.prunerStorage = prunerStorage;
+    this.blocksToRetain = blocksToRetain;
+    this.pruningFrequency = pruningFrequency;
+  }
+
+  public static void enablePruning() {
+    pruningEnabled = true;
+  }
+
+  public static boolean isPruningEnabled() {
+    return pruningEnabled;
+  }
+
+  @Override
+  public void onBlockAdded(final BlockAddedEvent event) {
+    LOG.debug("New block added event: " + event);
+    // Get pruning mark
+    long blockNumber = event.getBlock().getHeader().getNumber();
+    Optional<Long> maybePruningMark = getPruningMark();
+    if (maybePruningMark.isEmpty()) {
+      // Set initial pruning mark
+      maybePruningMark = Optional.of(blockNumber);
+    }
+    long pruningMark = maybePruningMark.get();
+    if (blockNumber < pruningMark) {
+      // Ignore and warn if block number < pruning mark, this normally indicates the blocksToKeep is
+      // too small.
+      LOG.warn(
+          "Block added event: "
+              + event
+              + " has a block number of "
+              + blockNumber
+              + " < pruning mark "
+              + pruningMark);
+      return;
+    }
+    // Append block into fork blocks.
+    KeyValueStorageTransaction tx = prunerStorage.startTransaction();
+    Collection<Hash> forkBlocks = getForkBlocks(blockNumber);
+    forkBlocks.add(event.getBlock().getHash());
+    setForkBlocks(tx, blockNumber, forkBlocks);
+    // If a block is a new canonical head, start pruning.
+    if (event.isNewCanonicalHead()
+        && blockNumber - blocksToRetain - pruningMark >= pruningFrequency) {
+      while (blockNumber - pruningMark >= blocksToRetain) {
+        LOG.debug("Pruning chain data at pruning mark: " + pruningMark);
+        // Get a collection of old fork blocks that need to be pruned.
+        Collection<Hash> oldForkBlocks = getForkBlocks(pruningMark);
+        BlockchainStorage.Updater updater = blockchainStorage.updater();
+        for (Hash toPrune : oldForkBlocks) {
+          Optional<BlockBody> maybeBody = blockchainStorage.getBlockBody(toPrune);
+          if (maybeBody.isEmpty()) {
+            continue;
```

Comment:
It could be possible, though maybe not likely that there is a block header, transaction receipts etc. might still need to be pruned. A safer approach could be to just not remove transaction locations if we can't get the body as don't have the detail to find them.

Reply:
updated so that tx location gets pruned the last